File: ueueo-framework/ueueo-eventbus/src/main/java/com/ueueo/eventbus/distributed/OutboxSender.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Microsoft.Extensions.Options;
using Volo.Abp.DependencyInjection;
using Volo.Abp.DistributedLocking;
using Volo.Abp.EventBus.Distributed;
using Volo.Abp.Threading;

namespace Volo.Abp.EventBus.Boxes;

public class OutboxSender : IOutboxSender, ITransientDependency
{
    protected IServiceProvider ServiceProvider;//  { get; }
    protected AbpAsyncTimer Timer;//  { get; }
    protected IDistributedEventBus DistributedEventBus;//  { get; }
    protected IAbpDistributedLock DistributedLock;//  { get; }
    protected IEventOutbox Outbox ;// { get; private set; }
    protected OutboxConfig OutboxConfig ;// { get; private set; }
    protected AbpEventBusBoxesOptions EventBusBoxesOptions;//  { get; }
    protected String DistributedLockName => "AbpOutbox_" + OutboxConfig.Name;
    public ILogger<OutboxSender> Logger;// { get; set; }

    protected CancellationTokenSource StoppingTokenSource;//  { get; }
    protected CancellationToken StoppingToken;//  { get; }

    public OutboxSender(
        IServiceProvider serviceProvider,
        AbpAsyncTimer timer,
        IDistributedEventBus distributedEventBus,
        IAbpDistributedLock distributedLock,
       IOptions<AbpEventBusBoxesOptions> eventBusBoxesOptions)
    {
        ServiceProvider = serviceProvider;
        DistributedEventBus = distributedEventBus;
        DistributedLock = distributedLock;
        EventBusBoxesOptions = eventBusBoxesOptions.Value;
        Timer = timer;
        Timer.Period = Convert.ToInt32(EventBusBoxesOptions.PeriodTimeSpan.TotalMilliseconds);
        Timer.Elapsed += TimerOnElapsed;
        Logger = NullLogger<OutboxSender>.Instance;
        StoppingTokenSource = new CancellationTokenSource();
        StoppingToken = StoppingTokenSource.Token;
    }

    public void StartAsync(OutboxConfig outboxConfig, CancellationToken cancellationToken = default)
    {
        OutboxConfig = outboxConfig;
        Outbox = (IEventOutbox)ServiceProvider.GetRequiredService(outboxConfig.ImplementationType);
        Timer.Start(cancellationToken);
        return Task.CompletedTask;
    }

    public void StopAsync(CancellationToken cancellationToken = default)
    {
        StoppingTokenSource.Cancel();
        Timer.Stop(cancellationToken);
        StoppingTokenSource.Dispose();
        return Task.CompletedTask;
    }

    private void TimerOnElapsed(AbpAsyncTimer arg)
    {
        RunAsync();
    }

    protected   void RunAsync()
    {
        using (var handle = DistributedLock.TryAcquireAsync(DistributedLockName, cancellationToken: StoppingToken))
        {
            if (handle != null)
            {
                while (true)
                {
                    var waitingEvents = Outbox.GetWaitingEventsAsync(EventBusBoxesOptions.OutboxWaitingEventMaxCount, StoppingToken);
                    if (waitingEvents.Count <= 0)
                    {
                        break;
                    }

                    Logger.LogInformation($"Found {waitingEvents.Count} events in the outbox.");

                    if (EventBusBoxesOptions.BatchPublishOutboxEvents)
                    {
                        PublishOutgoingMessagesInBatchAsync(waitingEvents);
                    }
                    else
                    {
                        PublishOutgoingMessagesAsync(waitingEvents);
                    }
                }
            }
            else
            {
                Logger.LogDebug("Could not obtain the distributed lock: " + DistributedLockName);
                try
                {
                    Task.Delay(EventBusBoxesOptions.DistributedLockWaitDuration, StoppingToken);
                }
                catch (TaskCanceledException) { }
            }
        }
    }

    protected   void PublishOutgoingMessagesAsync(List<OutgoingEventInfo> waitingEvents)
    {
        for (var waitingEvent in waitingEvents)
        {
            DistributedEventBus
                .AsSupportsEventBoxes()
                .PublishFromOutboxAsync(
                    waitingEvent,
                    OutboxConfig
                );

            Outbox.DeleteAsync(waitingEvent.Id);

            Logger.LogInformation($"Sent the event to the message broker with id = {waitingEvent.Id:N}");
        }
    }

    protected   void PublishOutgoingMessagesInBatchAsync(List<OutgoingEventInfo> waitingEvents)
    {
        DistributedEventBus
            .AsSupportsEventBoxes()
            .PublishManyFromOutboxAsync(waitingEvents, OutboxConfig);

        Outbox.DeleteManyAsync(waitingEvents.Select(x => x.Id).ToArray());

        Logger.LogInformation($"Sent {waitingEvents.Count} events to message broker");
    }
}

File: ueueo-framework/ueueo-eventbus/src/main/java/com/ueueo/eventbus/distributed/OutboxSender.java
package com.ueueo.eventbus.distributed;

import com.ueueo.distributedlocking.IDistributedLock;
import com.ueueo.distributedlocking.IDistributedLockHandle;
import com.ueueo.threading.CancellationToken;
import com.ueueo.uow.IUnitOfWorkManager;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.BeanFactory;
import org.springframework.beans.factory.config.ConfigurableBeanFactory;
import org.springframework.context.annotation.Scope;
import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
@Slf4j
public class OutboxSender implements IOutboxSender {
    protected BeanFactory beanFactory;
    private ThreadPoolTaskScheduler timer;
    protected IDistributedEventBus distributedEventBus;
    protected IDistributedLock distributedLock;
    protected IEventOutbox outbox;
    protected OutboxConfig outboxConfig;
    protected AbpEventBusBoxesOptions eventBusBoxesOptions;
    protected IUnitOfWorkManager unitOfWorkManager;
    protected CancellationToken stoppingToken;

    public OutboxSender(
            BeanFactory beanFactory,
            ThreadPoolTaskScheduler timer,
            IDistributedEventBus distributedEventBus,
            IDistributedLock distributedLock,
            IUnitOfWorkManager unitOfWorkManager,
            AbpEventBusBoxesOptions eventBusBoxesOptions,
            CancellationToken stoppingToken) {
        this.beanFactory = beanFactory;
        this.timer = timer;
        this.distributedEventBus = distributedEventBus;
        this.distributedLock = distributedLock;
        this.unitOfWorkManager = unitOfWorkManager;
        this.eventBusBoxesOptions = eventBusBoxesOptions;
        this.stoppingToken = stoppingToken;
    }

    @Override
    public void start(OutboxConfig outboxConfig) {
        this.outboxConfig = outboxConfig;
        outbox = (IEventOutbox) beanFactory.getBean(outboxConfig.getImplementationType());
        timer.scheduleAtFixedRate(this::run, eventBusBoxesOptions.getPeriodTimeSpan().toMillis());
    }

    @Override
    public void stop() {
        stoppingToken.cancel();
        timer.shutdown();
    }

    public String getDistributedLockName() {
        return "AbpOutbox_" + outboxConfig.getName();
    }

    protected void run() {
        if (stoppingToken.isCancellationRequested()) {
            return;
        }

        IDistributedLockHandle handle = distributedLock.tryAcquire(getDistributedLockName(), eventBusBoxesOptions.getDistributedLockWaitDuration());
        if (handle != null) {
            while (true) {
                List<OutgoingEventInfo> waitingEvents = outbox.getWaitingEvents(eventBusBoxesOptions.getOutboxWaitingEventMaxCount(), stoppingToken);
                if (waitingEvents.size() <= 0) {
                    break;
                }

                log.info("Found {} events in the outbox.", waitingEvents.size());

                if (eventBusBoxesOptions.isBatchPublishOutboxEvents()) {
                    publishOutgoingMessagesInBatch(waitingEvents);
                } else {
                    publishOutgoingMessages(waitingEvents);
                }
            }
            handle.dispose();
        } else {
            log.debug("Could not obtain the distributed lock: " + getDistributedLockName());
            try {
                Thread.sleep(eventBusBoxesOptions.getDistributedLockWaitDuration().toMillis());
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    protected void publishOutgoingMessages(List<OutgoingEventInfo> waitingEvents) {
        for (OutgoingEventInfo waitingEvent : waitingEvents) {
            distributedEventBus
                    .asSupportsEventBoxes()
                    .publishFromOutbox(waitingEvent, outboxConfig);

            outbox.delete(waitingEvent.getId());

            log.info("Sent the event to the message broker with id = {}", waitingEvent.getId().toString());
        }
    }

    protected void publishOutgoingMessagesInBatch(List<OutgoingEventInfo> waitingEvents) {
        distributedEventBus
                .asSupportsEventBoxes()
                .publishManyFromOutbox(waitingEvents, outboxConfig);

        outbox.deleteMany(waitingEvents.stream().map(OutgoingEventInfo::getId).collect(Collectors.toList()));

        log.info("Sent {} events to message broker", waitingEvents.size());
    }

}
